File: server/cpp/http.cpp
#include <boost/beast/core.hpp>
#include <boost/beast/http.hpp>
#include <boost/beast/version.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/filesystem.hpp>
#include <iostream>
#include <string>

namespace http = boost::beast::http;
namespace fs = boost::filesystem;

// 处理HTTP请求
void handle_request(http::request<http::string_body> req, http::response<http::string_body> &res)
{
    // 禁止POST请求
    if (req.method() != http::verb::get)
    {
        res.result(http::status::method_not_allowed);
        res.set(http::field::content_type, "text/plain");
        res.body() = "Method Not Allowed";
        return;
    }

    std::string path = "."; // 服务器根目录
    if (req.target() == "/")
    {
        path += "/index.html";
    }
    else
    {
        path += req.target().to_string();
    }

    try
    {
        if (fs::is_directory(path))
        {
            res.result(http::status::forbidden);
            res.set(http::field::content_type, "text/plain");
            res.body() = "Forbidden";
            return;
        }

        std::ifstream is(path, std::ios::in | std::ios::binary);
        if (!is)
        {
            res.result(http::status::not_found);
            res.set(http::field::content_type, "text/plain");
            res.body() = "Not Found";
            return;
        }

        // Read the file content into the response body
        std::ostringstream ostr;
        ostr << is.rdbuf();
        res.body() = ostr.str();

        // Set MIME type based on file extension
        std::string ext = fs::extension(path);
        if (ext == ".html" || ext == ".htm")
            res.set(http::field::content_type, "text/html");
        else if (ext == ".css")
            res.set(http::field::content_type, "text/css");
        else if (ext == ".js")
            res.set(http::field::content_type, "application/javascript");
        else if (ext == ".json")
            res.set(http::field::content_type, "application/json");
        else if (ext == ".jpg" || ext == ".jpeg")
            res.set(http::field::content_type, "image/jpeg");
        else if (ext == ".png")
            res.set(http::field::content_type, "image/png");
        else if (ext == ".svg")
            res.set(http::field::content_type, "image/svg+xml");
        else if (ext == ".ico")
            res.set(http::field::content_type, "image/x-icon");
        else if (ext == ".txt")
            res.set(http::field::content_type, "text/plain");
        else
            res.set(http::field::content_type, "application/octet-stream");

        res.result(http::status::ok);
    }
    catch (const std::exception &e)
    {
        res.result(http::status::internal_server_error);
        res.set(http::field::content_type, "text/plain");
        res.body() = "Internal Server Error";
    }
}

// 主函数
int main()
{
    namespace ip = boost::asio::ip;
    using tcp = ip::tcp;

    try
    {
        boost::asio::io_context ioc{1};
        tcp::acceptor acceptor{ioc, {tcp::v4(), 3000}};

        for (;;)
        {
            tcp::socket socket{ioc};
            acceptor.accept(socket);

            http::request<http::string_body> req;
            http::response<http::string_body> res;

            // Parse the request
            http::read(socket, boost::beast::basic_flat_buffer(), req);

            // Process the request and prepare the response
            handle_request(req, res);

            // Send the response
            http::write(socket, res);
        }
    }
    catch (const std::exception &e)
    {
        std::cerr << "Error: " << e.what() << "\n";
    }

    return 0;
}

File: server/nodejs/http.js
const http = require('http');
const fs = require('fs');
const path = require('path');

const port = 80;

// 定义支持的MIME类型
const mimes = {
    html: 'text/html',
    css: 'text/css',
    js: 'text/javascript',
    json: 'application/json',
    jpg: 'image/jpeg',
    png: 'image/png',
    svg: 'image/svg+xml',
    ico: 'image/x-icon',
    txt: 'text/plain'
}

// 创建HTTP服务器
const server = http.createServer((request, response) => {
    //禁止POST请求
    if (request.method !== 'GET') {
        response.writeHead(405, { 'Content-Type': 'text/plain' });
        response.write('Method Not Allowed');
        response.end();
        return;
    }
    let pathname = request.url; // 获取请求的路径
    console.log('Request for' + pathname);
    let root = path.join(__dirname, '../../'); // 设置根目录
    let filePath = root + pathname; // 拼接文件路径
    console.log('Request for' + filePath);
    if (pathname === '/') {
        filePath = root + 'index.html'; // 默认请求index.html
    }

    // 读取文件
    fs.readFile(filePath, (err, data) => {
        //错误处理
        if (err) {
            switch (err.code) {
                case 'ENOENT':
                    response.writeHead(404, { 'Content-Type': 'text/plain' });
                    response.write('Not Found');
                    response.end();
                    break;

                case 'EISDIR':
                    response.writeHead(403, { 'Content-Type': 'text/plain' });
                    response.write('Forbidden');
                    response.end();
                    break;

                default:
                    response.writeHead(500, { 'Content-Type': 'text/plain' });
                    response.write('Internal Server Error');
                    response.end();
                    break;
            }
            return;
        };
        //设置mime类型
        let ext = path.extname(filePath).slice(1); // 获取文件扩展名
        // 根据文件扩展名获取对应的 MIME 类型
        // 如果找不到对应的 MIME 类型，则默认为 'application/octet-stream'
        let contentType = mimes[ext] || 'application/octet-stream';

        // responst.setHeader('Content-Type', contentType + '; charset=UTF-8');
        response.setHeader('Content-Type', contentType); // 设置响应头

        response.end(data); // 结束响应，发送数据
    });
});

// 监听3000端口
server.listen(port, () => {
    console.log('Server is running at http://localhost/'); // 服务器启动提示
});